### PR TITLE
Add initial support for b3-propagation headers.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Code extracted from https://github.com/census-instrumentation/opencensus-java/pull/606.

### CODE REVIEW

File: api/src/test/java/io/opencensus/trace/SpanIdTest.java [hunk 1]
```diff
@@ -28,7 +28,7 @@
 @RunWith(JUnit4.class)
 public class SpanIdTest {
   private static final byte[] firstBytes = new byte[] {0, 0, 0, 0, 0, 0, 0, 'a'};
-  private static final byte[] secondBytes = new byte[] {(byte) 0xFF, 0, 0, 0, 0, 0, 0, 0};
+  private static final byte[] secondBytes = new byte[] {(byte) 0xFF, 0, 0, 0, 0, 0, 0, 'A'};
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply api/src/test/java/io/opencensus/trace/SpanIdTest.java]: Thanks.

File: api/src/main/java/io/opencensus/trace/SpanId.java [hunk 2]
```diff
@@ -83,6 +83,20 @@ public static SpanId fromBytes(byte[] src, int srcOffset) {
     return new SpanId(bytes);
   }
 
+  /**
+   * Returns a {@code SpanId} built from a lowercase base16 representation.
+   *
+   * @param src the lowercase base16 representation.
+   * @return a {@code SpanId} built from a lowercase base16 representation.
+   * @throws NullPointerException if {@code src} is null.
```
Comment (HailongWen) [inline api/src/main/java/io/opencensus/trace/SpanId.java]: Maybe add `throws IllegalArgumentException` in javadoc as `TraceId.fromLowerBase16` does?

File: api/src/main/java/io/opencensus/trace/TraceId.java [hunk 3]
```diff
@@ -84,6 +84,21 @@ public static TraceId fromBytes(byte[] src, int srcOffset) {
     return new TraceId(bytes);
   }
 
+  /**
+   * Returns a {@code TraceId} built from a lowercase base16 representation.
+   *
+   * @param src the lowercase base16 representation.
+   * @return a {@code TraceId} built from a lowercase base16 representation.
+   * @throws NullPointerException if {@code src} is null.
+   * @throws IllegalArgumentException if {@code buffer.length} is not {@code 2 * TraceId.SIZE}.
```
Comment (HailongWen) [inline api/src/main/java/io/opencensus/trace/TraceId.java]: s/buffer/src

File: impl_core/src/test/java/io/opencensus/implcore/trace/propagation/B3FormatTest.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  public void parseMissingSpanId() throws SpanContextParseException {
+    Map<String, String> invalidHeaders = new HashMap<String, String>();
+    invalidHeaders.put(X_B3_TRACE_ID, TRACE_ID_BASE16);
+    thrown.expect(SpanContextParseException.class);
+    thrown.expectMessage("Missing X_B3_SPAN_ID.");
+    b3Format.extract(invalidHeaders, getter);
+  }
+
+  @Test
+  public void fields_list() throws SpanContextParseException {
```
Comment (sebright) [inline impl_core/src/test/java/io/opencensus/implcore/trace/propagation/B3FormatTest.java]: This test doesn't throw a parse exception.

File: impl_core/src/main/java/io/opencensus/implcore/trace/propagation/B3Format.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+final class B3Format extends TextFormat {
+  @VisibleForTesting static final String X_B3_TRACE_ID = "X─B3─TraceId";
+  @VisibleForTesting static final String X_B3_SPAN_ID = "X─B3─SpanId";
+  @VisibleForTesting static final String X_B3_PARENT_SPAN_ID = "X─B3─ParentSpanId";
+  @VisibleForTesting static final String X_B3_SAMPLED = "X─B3─Sampled";
+  @VisibleForTesting static final String X_B3_FLAGS = "X-B3-Flags";
+  private static final List<String> FIELDS =
+      Collections.unmodifiableList(
+          Arrays.asList(
+              X_B3_TRACE_ID, X_B3_SPAN_ID, X_B3_PARENT_SPAN_ID, X_B3_SAMPLED, X_B3_FLAGS));
```
Comment (sebright) [inline impl_core/src/main/java/io/opencensus/implcore/trace/propagation/B3Format.java]: Why is X_B3_PARENT_SPAN_ID included in the list of fields if it isn't used by inject or extract?  Will it be used later?

File: impl_core/src/main/java/io/opencensus/implcore/trace/propagation/B3Format.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+  // Used as the upper 8-bytes of the traceId when a 8-bytes traceId is received.
+  private static final String UPPER_TRACE_ID = "0000000000000000";
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply impl_core/src/main/java/io/opencensus/implcore/trace/propagation/B3Format.java]: Tried something.

**Pull request merged**